### PR TITLE
Separate data loaders into GraphQL.DataLoader project

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

**Comment by Shane32** [comment]
Notes:
* The `IDataLoaderResult` and `IDataLoaderResult<T>` interfaces were not moved to the new project.  The former is necessary for the execution strategy.  The latter is not (at this time).

Those interfaces consist of these lines of code:
```cs
    public interface IDataLoaderResult<T> : IDataLoaderResult
    {
        new Task<T> GetResultAsync(CancellationToken cancellationToken = default);
    }

    public interface IDataLoaderResult
    {
        Task<object> GetResultAsync(CancellationToken cancellationToken = default);
    }
```

* All of the data loader extension methods were moved into the data loader project.  Some of them only rely on the `IDataLoaderResult<T>` interface and assist with the field builders.  I moved them anyway.

File: docs2/site/docs/guides/dataloader.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # DataLoader
 
-GraphQL.NET includes an implementation of Facebook's [DataLoader](https://github.com/facebook/dataloader).
+GraphQL.NET includes an implementation of Facebook's [DataLoader](https://github.com/facebook/dataloader) within the
+[`GraphQL.DataLoader`](https://www.nuget.org/packages/GraphQL.DataLoader/) NuGet package.
```
Comment (sungam3r) [inline docs2/site/docs/guides/dataloader.md:4]:
```suggestion
[`GraphQL.DataLoader`](https://www.nuget.org/packages/GraphQL.DataLoader) NuGet package.
```

File: docs2/site/docs/migrations/migration4.md [hunk 2]
```diff
@@ -254,6 +254,12 @@ protected override IExecutionStrategy SelectExecutionStrategy(ExecutionContext c
 }
 ```
 
+### DataLoader Moved to Separate Project
+
+The implementation for data loaders, contained within the `GraphQL.DataLoader` namespace, has been moved into the
+[`GraphQL.DataLoader`](https://www.nuget.org/packages/GraphQL.DataLoader/) NuGet package. Please import the NuGet
```
Comment (sungam3r) [inline docs2/site/docs/migrations/migration4.md:260]:
```suggestion
[`GraphQL.DataLoader`](https://www.nuget.org/packages/GraphQL.DataLoader) NuGet package. Please import the NuGet
```

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <Description>GraphQL data loader implementation</Description>
```
Comment (sungam3r) [inline src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: `data loader` or `dataloader` ? 🤔

**Comment by sungam3r** [comment]
Remember to update `Upload coverage to codecov` step.

**Comment by Shane32** [comment]
> Remember to update `Upload coverage to codecov` step.

I checked just now but it looks like it's uploading reports from the test projects.  Since the test projects didn't change, no change should be necessary -- correct?

**Comment by Shane32** [comment]
Or is there something else to change?

**Comment by sungam3r** [comment]
The fourth package appeared so it's time to add a table with nuget packages in readme as in the server project.

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj (hunk 3 above)
Comment (Shane32) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: Whatever you think. 🤷‍♂️

**Comment by sungam3r** [comment]
For some reason I thought that a new test project appeared and you need to register there. OK.

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: https://github.com/graphql/dataloader

search `dataloader` -> 83 matches
search `data loader` -> 0 mathes

😄

Comment (Shane32) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: ok so dataloader then?

Comment (Shane32) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]:
```suggestion
    <Description>GraphQL dataloader implementation</Description>
```

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    <PackageTags>GraphQL;data loader</PackageTags>
```
Comment (Shane32) [inline src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:6]:
```suggestion
    <PackageTags>GraphQL;dataloader</PackageTags>
```

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj (hunk 3 above)
Comment (sungam3r) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: Yes, everywhere.

Comment (sungam3r) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]: Well, even `DataLoader`. It is in such a register that this name is used.

Comment (Shane32) [reply src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:5]:
```suggestion
    <Description>GraphQL DataLoader implementation</Description>
```

**Review by sungam3r — APPROVED** [review]
Remember to change `dataloader` and `data loader` to `DataLoader`.

File: src/GraphQL.DataLoader/GraphQL.DataLoader.csproj [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <Description>GraphQL DataLoader implementation</Description>
+    <PackageTags>GraphQL;dataloader</PackageTags>
```
Comment (Shane32) [inline src/GraphQL.DataLoader/GraphQL.DataLoader.csproj:6]:
```suggestion
    <PackageTags>GraphQL;DataLoader</PackageTags>
```

File: docs2/site/docs/migrations/migration4.md [hunk 6]
```diff
@@ -254,6 +254,12 @@ protected override IExecutionStrategy SelectExecutionStrategy(ExecutionContext c
 }
 ```
 
+### DataLoader Moved to Separate Project
+
+The implementation for data loaders, contained within the `GraphQL.DataLoader` namespace, has been moved into the
+[`GraphQL.DataLoader`](https://www.nuget.org/packages/GraphQL.DataLoader) NuGet package. Please import the NuGet
+package if you use data loaders. No code changes are necessary.
```
Comment (Shane32) [inline docs2/site/docs/migrations/migration4.md:261]: If you're referring to this, I'm using it as an adjective/common noun separately from a proper noun.  When used as a proper noun, it is capitalized as DataLoader, and when used as an adjective (data) and common noun (loader) it is written as "data loader".

I did change the capitalization on the package tags as requested.

I see you approved the PR so I'll merge it.  It is easy enough to change the documentation if you think we should.

Comment (sungam3r) [reply docs2/site/docs/migrations/migration4.md:261]: Ok

**Pull request merged**